Support preserving unknown fields in ProtoBuf format

xiaozhikang0916 opened the pull request:
Introducing new annotation `@ProtoUnknownFields` and built-in type `ProtoMessage`, to store any fields with proto ids that are undefined in current message class.

In my understanding, both ktx-serialization and protocol buffer are designed in strong typing, all messages used in code should be well defined in kt class or proto message. Therefore I would like to keep `ProtoMessage` and `ProtoField` minimun published. Not like `JsonObject` in json format, users cannot retrieve detailed data from it.

Close #2655

xiaozhikang0916 committed:
introduce ProtoUnknownFields

xiaozhikang0916 committed:
add some test

xiaozhikang0916 committed:
comments

xiaozhikang0916 committed:
docs

xiaozhikang0916 committed:
api dump

xiaozhikang0916 commented:
@sandwwraith Hi, please take a look here?

xiaozhikang0916 committed:
add unit test for unknown one-of field

sandwwraith commented:
@xiaozhikang0916 Sure, when I'll have time

pdvrieze reviewed (commented):
The structure you use to represent protobuf content is somewhat complex, especially considering the protobuf wire format. You probably want to think a bit more how to represent protobuf data where the wire based information permits that. In particular you may want to distinguish primitives, "blobs", packed arrays. You may also allow users to observe a protobuf with provided metadata.

xiaozhikang0916 commented:
@pdvrieze Actually I don't want to decode the protobuf wire data into a tree-structure type like what `JsonObject` do, neither provide retreiving method of the content.

The raw content in protobuf wire data is quite ambiguous that, you can even not possible to tell whether a `VARINT 1` is 1 or -1 or true, or tell a `LEN` field is string or sub-message. The only way to parse wire data content correctly and explictly is to parse with message definition, or with a concrete type in kotlin.

Base on it, `ProtoMessage` here is not designed for retrieving content from undefined fields, but for holding it only.

xiaozhikang0916 commented:
Any progress about it? We have plan on upgrading Kotlin and serialization runtime version, and it would be great if we can have this functionality in near future.

shanshin commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt:
```diff
@@ -69,21 +75,34 @@ internal open class ProtobufDecoder(
+                }
+                ID_HOLDER_UNKNOWN_FIELDS -> {
+                    require(unknownHolderIndex == -1) {
+                        "Only one unknown fields holder is allowed in a message"
```
Should we write descriptor name here?

shanshin commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt:
```diff
@@ -69,21 +75,34 @@ internal open class ProtobufDecoder(
+                        "Only one unknown fields holder is allowed in a message"
+                    }
+                    require(descriptor.getElementDescriptor(i).nullable == ProtoMessageSerializer.descriptor.nullable) {
+                        "ProtoUnknownFields is only allowed on property with type kotlinx.serialization.protobuf.ProtoMessage, with its original serializer."
```
And here

shanshin commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoMessage.kt:
```diff
@@ -0,0 +1,199 @@
+    /**
+     * Returns a byte array representing of the message.
+     */
+    public fun asByteArray(): ByteArray =
```
The current implementation looks like it overload the GC, but it is not used anywhere.
I would prefer to delete it in this form.

shanshin commented on formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/ProtobufUnknownFieldsTest.kt:
```diff
@@ -0,0 +1,362 @@
+         *   3: {"42"}
+         * }
+         */
+        val encoded = "082a120234321a032a2a2a202a202a202a2a120a023432102a1a0234321a0234321a023432"
```
I would suggest building the main tests not around the hardcoded HEX values, but according to the following rule: everything that is encoded must be decoded in exactly the same form.

Steps:
- first encode what is
- then decode it to type with `ProtoUnknownFields`
- compare known fields
- encode the type with `ProtoUnknownFields` 
- this value must match the first encoded one

This is almost the same approach, but it will reduce the use of magic constants in all tests.

Of course, it is possible to check decoding from a constant HEX value, but one test will be enough for this.

xiaozhikang0916 committed:
Fix proto unknown fields reading

xiaozhikang0916 committed:
Introduce INDEX_NOT_EXISTSED for init or default value of index.

xiaozhikang0916 committed:
improve invalid unknown fields message

xiaozhikang0916 committed:
Rename ProtoMessage to ProtoUnknownFieldHolder

xiaozhikang0916 committed:
Update api dump and docs

xiaozhikang0916 commented on docs/formats.md:
```diff
@@ -642,6 +643,54 @@ is also compatible with the `message Data` given above, which means the same inp
+
+Kotlin Serialization `ProtoBuf` format supports preserving unknown fields, as described in the [Protocol Buffer-Unknown Fields](https://protobuf.dev/programming-guides/proto3/#unknowns).
+
+To keep the unknown fields, add a property in type `ProtoMessage` with default value `null` or `ProtoMessage.Empty`, and annotation `@ProtoUnknownFields` to your data class.
```
`ProtoContentHolder` is also removed.

xiaozhikang0916 committed:
simplify raw element encoding and decoding

xiaozhikang0916 committed:
Add comprehensive unknown fields round-trip tests

Add tests for all protobuf wire types (varint, fixed32, fixed64, size-delimited)

xiaozhikang0916 committed:
Rewrite unknown field test to encode-decode pattern

sandwwraith commented:
Did we discuss an overall design for the feature? I have three questions:

1. If we support only `ProtoUnknownFieldHolder` type, the annotation seems unnecessary. We have full access to the descriptor everywhere, so we can simply check descriptor/serialName for equality.

2. Do we want to support only this type? It is typical to store unknown fields only in `ByteArray` — in case we are not interested in them at all, but want to preserve them when we'll serialize the message again later. Why choose a separate type? Especially given that `ProtoField` is internal, there's no way to access at least protonumbers of new fields (in case you want to, e.g., log or dump them).  `ProtoUnknownFieldHolder` probably makes more sense when ProtoField is public. But ProtoField contains a ByteArray anyway, so we're back to square one. 
3. Maybe we need functionality to parse a raw ByteArray to raw unknown fields? Then we can get in the cycle ProtoUnknownFieldHolder - ProtoField - ByteArray - inner message as ProtoUnknownFieldHolder - ProtoField - ....

xiaozhikang0916 committed:
Remove ProtoUnknownFields annotation

xiaozhikang0916 committed:
Update docs apis and error messages for unknown fields

xiaozhikang0916 committed:
Simplify ProtoUnknownFieldHolder with byteArray as content

xiaozhikang0916 committed:
Update api for ProtoUnknownFieldHolder

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/Helpers.kt:
```diff
@@ -46,16 +46,27 @@ internal enum class ProtoWireType(val typeId: Int) {
 internal const val ID_HOLDER_ONE_OF = -2
+internal const val ID_HOLDER_UNKNOWN_FIELDS = -3
 
+internal const val INDEX_NOT_EXISTSED = -1
```
Typo, should be EXISTED

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt:
```diff
@@ -69,21 +75,31 @@ internal open class ProtobufDecoder(
+                }
+                ID_HOLDER_UNKNOWN_FIELDS -> {
+                    require(unknownHolderIndex == INDEX_NOT_EXISTSED) {
+                        "Only one unknown fields holder is allowed in a message, but get ${descriptor.getElementName(i)} and ${descriptor.getElementName(unknownHolderIndex)}"
```
```suggestion
                        "Only one unknown fields holder is allowed in a message, but two properties have ProtoUnknownFieldHolder type: ${descriptor.getElementName(i)} and ${descriptor.getElementName(unknownHolderIndex)}"
```

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt:
```diff
@@ -254,6 +270,9 @@ internal open class ProtobufDecoder(
             deserializer.descriptor == UByteArraySerializer().descriptor -> deserializeByteArray((previousValue as UByteArray?)?.asByteArray()).asUByteArray() as T
             deserializer is AbstractCollectionSerializer<*, *, *> ->
                 (deserializer as AbstractCollectionSerializer<*, T, *>).merge(this, previousValue)
+            deserializer == ProtoUnknownFieldHolderSerializer -> {
```
You should compare `deserializer.descriptor == ProtoUnknownFieldHolderSerializer.descriptor` — see how it is done with `deserializeByteArray` above

xiaozhikang0916 replied:
Changed, but can you elaborate why this pattern is recommended here?

sandwwraith replied:
We do not require any equality contract from serializers, and keep the instantiation strategy an implementation detail. Changing `object` to `class` and vice versa may break the serializer's identity, while `SerialDescriptor.equals` is well-defined.

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufEncoding.kt:
```diff
@@ -143,11 +143,18 @@ internal open class ProtobufEncoder(
         serializer is MapLikeSerializer<*, *, *, *> -> {
             serializeMap(serializer as SerializationStrategy<T>, value)
         }
+        serializer == ProtoUnknownFieldHolderSerializer -> {
```
Same here

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtoUnknownFieldHolderSerializer.kt:
```diff
@@ -0,0 +1,79 @@
+    if (compositeDecoder is ProtobufDecoder) {
+        return readRawFieldBytes(compositeDecoder, compositeDecoder.currentTag)
+    }
+    return ByteArray(0)
```
I think it's better to fail with ClassCastException to ProtobufDecoder rather than silently return wrong data

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtoUnknownFieldHolderSerializer.kt:
```diff
@@ -0,0 +1,79 @@
+    return ByteArray(0)
+}
+
+internal class UnknownFieldsDescriptor(private val original: SerialDescriptor) : SerialDescriptor by original {
```
I don't see the point of making a wrapper class if the only instance will be in `ProtoUnknownFieldHolderSerializer`

xiaozhikang0916 replied:
If we check whether a deserializer is `ProtoUnknownFieldHolderSerializer` by comparing its descriptor, a wrapper is needed here to do the comparasion, otherwise, a `ByteArraySerializer` may be considered same type as `ProtoUnknownFieldHolderSerializer` in `deserializer.descriptor == ProtoUnknownFieldHolderSerializer.descriptor`.

sandwwraith replied:
`SerialDescriptor("kotlinx.serialization.protobuf.ProtoUnknownFieldHolder", ByteArraySerializer().descriptor)` provides such wrapper

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtoUnknownFieldHolderSerializer.kt:
```diff
@@ -0,0 +1,79 @@
+import kotlinx.serialization.protobuf.*
+
+internal object ProtoUnknownFieldHolderSerializer : KSerializer<ProtoUnknownFieldHolder> {
+    override val descriptor: SerialDescriptor
```
You don't really need to create new descriptor every time via `get()`. You can use a specialized [delegate descriptor constructor](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization.descriptors/-serial-descriptor.html): `override val descriptor: SerialDescriptor = SerialDescriptor("kotlinx.serialization.protobuf.ProtoUnknownFieldHolder", ByteArraySerializer().descriptor)`

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoUnknownFieldHolder.kt:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Represents a protobuf message.
+ *
+ * Especially used as a holder of unknown proto fields in an arbitrary protobuf message.
```
Since representing unknown fields is its only usage, we should rephrase this. Maybe

`Represents a thin wrapper over fields in a protobuf message. Used to store unknown proto fields.`

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoUnknownFieldHolder.kt:
```diff
@@ -0,0 +1,56 @@
+ */
+@Serializable(with = ProtoUnknownFieldHolderSerializer::class)
+public class ProtoUnknownFieldHolder internal constructor(
+    internal val fields: ByteArray
```
Should we make it public? At least priniting them is probably a popular request. One may also want to try to edit them before writing back

xiaozhikang0916 replied:
Sure. And I believe a list of fields data is more helpful than concating them into one byte array so that developers can know how many unknown fields is captured in this holder.

> One may also want to try to edit them before writing back

It is quite diffcult to do that in this implementation, because it is hard to know the id of these fields. To store and retrieve the id, we need a holder for arbitrary proto fields, but we just desided to remove `ProtoField` introduced in this pr.

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoUnknownFieldHolder.kt:
```diff
@@ -0,0 +1,56 @@
+        return fields.contentHashCode()
+    }
+
+    override fun equals(other: Any?): Boolean {
```
`toString()` is missing

sandwwraith commented on formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/ProtobufUnknownFieldsTest.kt:
```diff
@@ -0,0 +1,496 @@
+    fun testUnknownFloat() {
+        assertUnknownFieldRoundTrip(FloatData(1, 0.0f), FloatData.serializer())
+        assertUnknownFieldRoundTrip(FloatData(1, -1.0f), FloatData.serializer())
+        assertUnknownFieldRoundTrip(FloatData(1, Float.MIN_VALUE), FloatData.serializer())
```
This test fails on JS, probably due to a difference in precision. Can you investigate whether it is safe to ignore this failure?

sandwwraith replied:
`Expected <FloatData(id=1, value=1.4e-45)>, actual <FloatData(id=1, value=1.401298464324817e-45)>`

xiaozhikang0916 replied:
I believe it is save to ignore this failure here. Json and Cbor only test float value in jvm target because of this precision issue.

xiaozhikang0916 committed:
[proto] rename typo

xiaozhikang0916 committed:
[proto] update duplicate ProtoUnknownFieldHolder error message

xiaozhikang0916 committed:
[proto] update ProtoUnknownFieldHolderSerializer instance check

xiaozhikang0916 committed:
[proto] strick check while reading raw field data

xiaozhikang0916 committed:
[proto] delegate serial descriptor of ProtoUnknownFieldHolderSerializer

xiaozhikang0916 committed:
[proto] Desc of ProtoUnknownFieldHolder

xiaozhikang0916 committed:
[proto] added toString to ProtoUnknownFieldHolder with the only public size property.

xiaozhikang0916 committed:
[proto] Only test special float value in jvm

xiaozhikang0916 committed:
[proto] remove unused test class

xiaozhikang0916 committed:
[proto] Make fields in ProtoUnknownFieldHolder a public list

xiaozhikang0916 committed:
[proto] remove UnknownFieldsDescriptor

sandwwraith commented on docs/formats.md:
```diff
@@ -666,6 +643,54 @@ is also compatible with the `message Data` given above, which means the same inp
+
+Kotlin Serialization `ProtoBuf` format supports preserving unknown fields, as described in the [Protocol Buffer-Unknown Fields](https://protobuf.dev/programming-guides/proto3/#unknowns).
+
+To keep the unknown fields, add a property in type `ProtoUnknownFieldHolder` with default value `null` or `ProtoUnknownFieldHolder.Empty` to your data class.
```
```suggestion
To keep the unknown fields, add a property of type `ProtoUnknownFieldHolder` with default value `null` or `ProtoUnknownFieldHolder.Empty` to your data class.
```

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoUnknownFieldHolder.kt:
```diff
@@ -0,0 +1,58 @@
+ */
+@Serializable(with = ProtoUnknownFieldHolderSerializer::class)
+public class ProtoUnknownFieldHolder internal constructor(
+    public val fields: List<ByteArray>
```
I'm still not convinced we should provide any API beyond simple `public val fields: ByteArray`. Can you provide an example where the number of such fields matters, and where having a `List` would make a big difference? IMO, it is enough to know just size and raw content for some toString/debuggability, and you can't really do anything beyond that.

If we are going to add `ProtoUnknownField` in the future, it will also be easier to hide plain byte array rather than `fields+contentSize+size` altogether.

xiaozhikang0916 replied:
Got it. Any further API for the internal fields will be suspended until a real `ProtoField` is introduced.

xiaozhikang0916 committed:
[proto] update format docs

xiaozhikang0916 committed:
Revert "[proto] Make fields in ProtoUnknownFieldHolder a public list"

This reverts commit 2aac4334493a77499aec58e7f86cfec495bf34ef.

xiaozhikang0916 committed:
[proto] make fields in ProtoUnknownFieldHolder public

sandwwraith commented on docs/formats.md:
```diff
@@ -666,6 +643,54 @@ is also compatible with the `message Data` given above, which means the same inp
+
+You may keep updating your schema by adding new fields, but you may not want to break compatibility with the old data.
+
+Kotlin Serialization `ProtoBuf` format supports preserving unknown fields, as described in the [Protocol Buffer-Unknown Fields](https://protobuf.dev/programming-guides/proto3/#unknowns).
```
```suggestion
Kotlin Serialization `ProtoBuf` format supports preserving unknown fields, as described in the [Protocol Buffer - Unknown Fields](https://protobuf.dev/programming-guides/proto3/#unknowns).
```

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/Helpers.kt:
```diff
@@ -133,9 +147,15 @@ internal fun extractProtoId(descriptor: SerialDescriptor, index: Int, zeroBasedD
             }
         }
     }
+    if (descriptor.getElementDescriptor(index).isUnknownFieldsDescriptor) {
```
Minor: KDoc for this function has to be updated as well.

Also, it looks suspiciously similar to `fun SerialDescriptor.extractParameters`. Later investigation is needed on whether we can unify them (out of scope for this PR)

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtoUnknownFieldHolderSerializer.kt:
```diff
@@ -0,0 +1,70 @@
+ * Reads the complete wire format bytes (tag + value) for the current field
+ * from [compositeDecoder] using its current tag.
+ */
+internal fun readRawFieldBytes(compositeDecoder: CompositeDecoder): ByteArray {
```
You can have `compositeDecoder: ProtobufDecoder` here because of smartcast. You don't need a second `is`.

xiaozhikang0916 replied:
`this` in the lambda of `Decoder.decodeStructure` has type in `CompositeDecoder`, which is not smartcasted to `ProtobufDecoder`. Do you suggest accessing `decoder` outside of the lambda scope here?

sandwwraith replied:
Hm, indeed. You can access `currentTag` anyway, so you can drop `internal fun readRawFieldBytes(compositeDecoder: CompositeDecoder)` completely, leaving only `readRawFieldBytes(decoder: ProtobufDecoder, currentTag: ProtoDesc)`.

Or you can edit ProtobufDecoder to narrow the return type on overload: `override fun beginStructure(descriptor: SerialDescriptor): ProtobufDecoder`. You'll have to inline decodeStructure lambda though.

xiaozhikang0916 replied:
Yeah, `internal fun readRawFieldBytes(compositeDecoder: CompositeDecoder)` and some unnecessary type cast have been removed in my last commit.

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtoUnknownFieldHolderSerializer.kt:
```diff
@@ -0,0 +1,70 @@
+    )
+
+    override fun deserialize(decoder: Decoder): ProtoUnknownFieldHolder {
+        if (decoder is ProtobufDecoder) {
```
Follow the pattern with `.asJsonEncoder/Decoder` (https://github.com/Kotlin/kotlinx.serialization/blob/304e858ccc7066854637d86ab80056f5f2bcc094/formats/json/commonMain/src/kotlinx/serialization/json/JsonElementSerializers.kt#L203) — perform check here, throw IllegalStateException on failure instead of CCE.

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtoUnknownFieldHolderSerializer.kt:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    override fun serialize(encoder: Encoder, value: ProtoUnknownFieldHolder) {
+        if (encoder is ProtobufEncoder) {
```
Same here — ISE on failed cast 

Alternatively, you can just call `encodeSerializableValue` with ByteArraySerializer if you need Json support for ProtoUnknownFieldHolder. But I'm not sure it is needed.

sandwwraith commented on formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtoUnknownFieldHolderSerializer.kt:
```diff
@@ -0,0 +1,70 @@
+ * Reads the complete wire format bytes (tag + value) for the current field
+ * from [decoder]. The [currentTag] provides the proto id for the unknown field.
+ */
+internal fun readRawFieldBytes(decoder: ProtobufDecoder, currentTag: ProtoDesc): ByteArray {
```
This whole method can be moved inside `decodeRawElement(protoId: Int)` since `decoder` already knows `currentType`

xiaozhikang0916 committed:
[proto] Fix title of links in docs

xiaozhikang0916 committed:
[proto] Update docs for extractProtoId

xiaozhikang0916 committed:
[proto] Update ProtoUnknownFieldHolderSerializer on type checking

sandwwraith reviewed (approved):
Thanks a lot for this huge missing piece in protobuf support!

xiaozhikang0916 commented:
What should I do to progress the related PR in Kotlin https://github.com/JetBrains/kotlin/pull/5755?
